File: python_solutions/1567.maximum-length-of-subarray-with-positive-product.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right 
from functools import reduce, lru_cache 
from typing import List 
import itertools 
import math 
import heapq 
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
#
# @lc app=leetcode id=1567 lang=python3
#
# [1567] Maximum Length of Subarray With Positive Product
#
# https://leetcode.com/problems/maximum-length-of-subarray-with-positive-product/description/
#
# algorithms
# Medium (27.12%)
# Total Accepted:    4.4K
# Total Submissions: 14.9K
# Testcase Example:  '[1,-2,-3,4]'
#
# Given an array of integers nums, find the maximum length of a subarray where
# the product of all its elements is positive.
# 
# A subarray of an array is a consecutive sequence of zero or more values taken
# out of that array.
# 
# Return the maximum length of a subarray with positive product.
# 
# 
# Example 1:
# 
# 
# Input: nums = [1,-2,-3,4]
# Output: 4
# Explanation: The array nums already has a positive product of 24.
# 
# 
# Example 2:
# 
# 
# Input: nums = [0,1,-2,-3,-4]
# Output: 3
# Explanation: The longest subarray with positive product is [1,-2,-3] which
# has a product of 6.
# Notice that we cannot include 0 in the subarray since that'll make the
# product 0 which is not positive.
# 
# Example 3:
# 
# 
# Input: nums = [-1,-2,-3,0,1]
# Output: 2
# Explanation: The longest subarray with positive product is [-1,-2] or
# [-2,-3].
# 
# 
# Example 4:
# 
# 
# Input: nums = [-1,2]
# Output: 1
# 
# 
# Example 5:
# 
# 
# Input: nums = [1,2,3,5,-6,4,0,10]
# Output: 4
# 
# 
# 
# Constraints:
# 
# 
# 1 <= nums.length <= 10^5
# -10^9 <= nums[i] <= 10^9
# 
# 
#
class Solution:
    def getMaxLen(self, nums: List[int]) -> int:
        def solve(i, j):
            neg_count = sum(1 if n < 0 else 0 for n in nums[i:j+1])
            if neg_count % 2 == 0: 
                return j - i + 1 
            x = next(itertools.dropwhile(lambda k: nums[k] > 0, range(i, j + 1)), j)
            y = next(itertools.dropwhile(lambda k: nums[k] > 0, range(j, i - 1, -1)), i)
            # print(neg_count, i, j, x, y)
            return max(y - i, j - x)

        i = j = 0
        res = 0 
        nums.append(0)
        while j < len(nums):
            # print(j)
            while nums[j] != 0 and j < len(nums): j += 1
            res = max(res, solve(i, j - 1))
            print(res, i, j)
            i = j + 1
            j += 1
        return res 

        

sol = Solution()


nums = [1,-2,-3,4]
nums = [0,1,-2,-3,-4]
nums = [-1,-2,-3,0,1]
nums = [-1,2]
nums = [1,2,3,5,-6,4,0,10]
print(sol.getMaxLen(nums))
# x = next(itertools.dropwhile(lambda i: nums[i] > 0, range(len(nums))))
# y = next(itertools.dropwhile(lambda i: nums[i] > 0, range(len(nums)-1, -1, -1)))
# print(x, y)

File: python_solutions/1567.maximum-length-of-subarray-with-positive-product.cpp
// C libraries
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>

// Containers
#include <deque>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// Input/Output
#include <fstream>
#include <iomanip>
#include <ios>
#include <iostream>
#include <istream>
#include <ostream>
#include <sstream>

// Other
#include <algorithm>
#include <bitset>
#include <chrono>
#include <iterator>
#include <limits>
#include <random>
#include <stdexcept>
#include <string>
#include <tuple>
#include <type_traits>
#include <utility>

// ==================================================

using namespace std;

// constants
const double EPS = 1e-9;
const int MOD = 1000000007;

// type alias
using pii = pair<int, int>;
using ll = long long;
using ull = unsigned long long;
using vi = vector<int>;
using vb = vector<bool>;
using vc = vector<char>;
using vs = vector<string>;
using vvb = vector<vector<bool>>;
using vvc = vector<vector<char>>;
using vvi = vector<vector<int>>;
using vvs = vector<vector<string>>;
using mii = map<int, int>;
using mci = map<char, int>;
using si = set<int>;
using spii = set<pair<int, int>>;
using umii = unordered_map<int, int>;
using umci = unordered_map<char, int>;
using umsi = unordered_map<string, int>;
using usi = unordered_set<int>;
using usc = unordered_set<char>;
using uss = unordered_set<string>;
using vpii = vector<pair<int, int>>;

typedef struct TreeNode TreeNode;
using ptn = TreeNode *;
using tn = TreeNode;

// ==================================================

// fast IO
static auto __speedup__ = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  return 0;
}();

// ==================================================

// some macro for less typing
#define forloop(i, n) for (int i = 0; i < n; i++)             //[0, n)
#define forloopr(i, n) for (int i = n - 1; i >= 0; --i)       // reverse [0, n)
#define forloopup(i, a, b) for (int i = a; i < b; ++i)        // [a, b)
#define forloopdown(i, a, b) for (int i = b - 1; i >= a; --i) // reverse [a, b)
#define forunfold(i, arr) for (auto &i : arr)

#define INF 0x3f3f3f3f
#define MAX(a, b) a = max(a, b)
#define MIN(a, b) a = min(a, b)

#define fi first
#define se second
#define mp make_pair
#define pb push_back
#define pf push_front
#define ef emplace_front
#define eb emplace_back
#define qall(v) v.begin(), v.end()
#define qsize(v) (int)v.size()
#define qsort(v) sort(v.begin(), v.end())
#define rqsort(v) sort(v.rbegin(), v.rend())
#define qreverse(v) reverse(v.begin(), v.end())

// int to string
string itos(int n) { return to_string(n); }
// char to string
string ctos(char c) { return string(1, c); };

inline string upper(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::toupper);
  return t;
}
inline string lower(string s) {
  string t(s);
  transform(t.begin(), t.end(), t.begin(), ::tolower);
  return t;
}

int dirs[5] = {-1, 0, 1, 0, -1};

// ==================================================
/*
 * @lc app=leetcode id=1567 lang=cpp
 *
 * [1567] Maximum Length of Subarray With Positive Product
 *
 * https://leetcode.com/problems/maximum-length-of-subarray-with-positive-product/description/
 *
 * algorithms
 * Medium (27.12%)
 * Total Accepted:    4.4K
 * Total Submissions: 14.9K
 * Testcase Example:  '[1,-2,-3,4]'
 *
 * Given an array of integers nums, find the maximum length of a subarray where
 * the product of all its elements is positive.
 *
 * A subarray of an array is a consecutive sequence of zero or more values
 * taken out of that array.
 *
 * Return the maximum length of a subarray with positive product.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [1,-2,-3,4]
 * Output: 4
 * Explanation: The array nums already has a positive product of 24.
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [0,1,-2,-3,-4]
 * Output: 3
 * Explanation: The longest subarray with positive product is [1,-2,-3] which
 * has a product of 6.
 * Notice that we cannot include 0 in the subarray since that'll make the
 * product 0 which is not positive.
 *
 * Example 3:
 *
 *
 * Input: nums = [-1,-2,-3,0,1]
 * Output: 2
 * Explanation: The longest subarray with positive product is [-1,-2] or
 * [-2,-3].
 *
 *
 * Example 4:
 *
 *
 * Input: nums = [-1,2]
 * Output: 1
 *
 *
 * Example 5:
 *
 *
 * Input: nums = [1,2,3,5,-6,4,0,10]
 * Output: 4
 *
 *
 *
 * Constraints:
 *
 *
 * 1 <= nums.length <= 10^5
 * -10^9 <= nums[i] <= 10^9
 *
 *
 */

class Solution {
public:
  int getMaxLen(vector<int> &nums) {
    nums.push_back(0);
    int res = 0, i = 0, j = 0, n = nums.size(), flag = 1;

    while (j < n) {
      if (nums[j] == 0) {
        if (flag == 1) res = max(j - i, res);
        else {
          while (i < j && nums[i] > 0) i++;
          res = max(j - i - 1, res);
        }
        flag = 1;
        i = j + 1;
      } else {
        flag = nums[j] < 0 ? -1 * flag : flag;
        if (flag == 1) res = max(j - i + 1, res);
      }
      // std::cout << j << ", " << res << ", " << flag << std::endl;
      // say(vector<int>{j, res, flag});
      j++;
    }
    return res;
  }
};

auto speed_up = []() {
  ios_base::sync_with_stdio(false);
  return 0;
}();
